File: merry360x_mobile/ios-app/Merry360xMobile/Views/Components/MerryLoadingAnimation.swift
import SwiftUI

struct MerryLoadingAnimation: View {
    @State private var activeIndex = 0
    @State private var scales: [CGFloat] = [1, 1, 1, 1]
    @State private var opacities: [Double] = [0.3, 0.3, 0.3, 0.3]
    
    private let icons = ["house.fill", "building.2.fill", "car.fill", "airplane"]
    private let iconColors: [Color] = [
        Color(red: 1, green: 0.34, blue: 0.25),   // Coral - House
        Color(red: 0.4, green: 0.6, blue: 1),     // Blue - Apartment
        Color(red: 0.3, green: 0.8, blue: 0.5),   // Green - Car
        Color(red: 1, green: 0.6, blue: 0.2)      // Orange - Airplane
    ]
    
    let timer = Timer.publish(every: 0.4, on: .main, in: .common).autoconnect()
    
    var body: some View {
        VStack(spacing: 24) {
            HStack(spacing: 20) {
                ForEach(0..<4, id: \.self) { index in
                    ZStack {
                        // Glow effect
                        Circle()
                            .fill(iconColors[index].opacity(0.3))
                            .frame(width: 50, height: 50)
                            .blur(radius: 10)
                            .scaleEffect(scales[index])
                            .opacity(opacities[index])
                        
                        // Icon
                        Image(systemName: icons[index])
                            .font(.system(size: 28, weight: .medium))
                            .foregroundColor(iconColors[index])
                            .scaleEffect(scales[index])
                            .opacity(opacities[index] + 0.4)
                    }
                    .frame(width: 50, height: 50)
                }
            }
            
            Text("Loading...")
                .font(.system(size: 14, weight: .medium))
                .foregroundColor(.gray)
        }
        .onReceive(timer) { _ in
            withAnimation(.easeInOut(duration: 0.3)) {
                // Reset previous
                let prevIndex = (activeIndex - 1 + 4) % 4
                scales[prevIndex] = 1
                opacities[prevIndex] = 0.3
                
                // Animate current
                scales[activeIndex] = 1.3
                opacities[activeIndex] = 1.0
                
                // Move to next
                activeIndex = (activeIndex + 1) % 4
            }
        }
    }
}

// Full screen loading overlay
struct MerryLoadingOverlay: View {
    let message: String
    
    init(message: String = "Loading...") {
        self.message = message
    }
    
    var body: some View {
        ZStack {
            Color.black.opacity(0.4)
                .ignoresSafeArea()
            
            VStack(spacing: 24) {
                MerryLoadingAnimation()
                
                Text(message)
                    .font(.system(size: 16, weight: .medium))
                    .foregroundColor(.white)
            }
            .padding(32)
            .background(
                RoundedRectangle(cornerRadius: 20)
                    .fill(.ultraThinMaterial)
            )
        }
    }
}

// Compact inline loading
struct MerryLoadingCompact: View {
    @State private var rotation: Double = 0
    
    private let icons = ["house.fill", "building.2.fill", "car.fill", "airplane"]
    private let iconColors: [Color] = [
        Color(red: 1, green: 0.34, blue: 0.25),
        Color(red: 0.4, green: 0.6, blue: 1),
        Color(red: 0.3, green: 0.8, blue: 0.5),
        Color(red: 1, green: 0.6, blue: 0.2)
    ]
    
    var body: some View {
        ZStack {
            ForEach(0..<4, id: \.self) { index in
                Image(systemName: icons[index])
                    .font(.system(size: 16, weight: .medium))
                    .foregroundColor(iconColors[index])
                    .offset(x: 18 * cos(Double(index) * .pi / 2 + rotation),
                            y: 18 * sin(Double(index) * .pi / 2 + rotation))
            }
        }
        .frame(width: 60, height: 60)
        .onAppear {
            withAnimation(.linear(duration: 2).repeatForever(autoreverses: false)) {
                rotation = .pi * 2
            }
        }
    }
}

// Bouncing dots style with travel icons
struct MerryLoadingBounce: View {
    @State private var bounceOffsets: [CGFloat] = [0, 0, 0, 0]
    
    private let icons = ["house.fill", "building.2.fill", "car.fill", "airplane"]
    private let iconColors: [Color] = [
        Color(red: 1, green: 0.34, blue: 0.25),
        Color(red: 0.4, green: 0.6, blue: 1),
        Color(red: 0.3, green: 0.8, blue: 0.5),
        Color(red: 1, green: 0.6, blue: 0.2)
    ]
    
    var body: some View {
        HStack(spacing: 12) {
            ForEach(0..<4, id: \.self) { index in
                Image(systemName: icons[index])
                    .font(.system(size: 24, weight: .medium))
                    .foregroundColor(iconColors[index])
                    .offset(y: bounceOffsets[index])
            }
        }
        .onAppear {
            for i in 0..<4 {
                withAnimation(
                    .easeInOut(duration: 0.5)
                    .repeatForever()
                    .delay(Double(i) * 0.15)
                ) {
                    bounceOffsets[i] = -15
                }
            }
        }
    }
}

#Preview("Standard") {
    MerryLoadingAnimation()
}

#Preview("Overlay") {
    ZStack {
        Color.gray.opacity(0.3)
        Text("Content behind")
    }
    .overlay(MerryLoadingOverlay(message: "Finding your stay..."))
}

#Preview("Compact") {
    MerryLoadingCompact()
}

#Preview("Bounce") {
    MerryLoadingBounce()
}

File: merry360x_mobile/android-app/app/src/main/java/com/merry360x/mobile/ui/components/MerryLoadingAnimation.kt
package com.merry360x.mobile.ui.components

import androidx.compose.animation.core.FastOutSlowInEasing
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.RepeatMode
import androidx.compose.animation.core.animateFloat
import androidx.compose.animation.core.infiniteRepeatable
import androidx.compose.animation.core.rememberInfiniteTransition
import androidx.compose.animation.core.tween
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.offset
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.AirplanemodeActive
import androidx.compose.material.icons.filled.Apartment
import androidx.compose.material.icons.filled.DirectionsCar
import androidx.compose.material.icons.filled.Home
import androidx.compose.material3.Icon
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.scale
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import kotlin.math.cos
import kotlin.math.sin

// Icon data
private data class LoadingIcon(
    val icon: ImageVector,
    val color: Color
)

private val loadingIcons = listOf(
    LoadingIcon(Icons.Default.Home, Color(0xFFFF5744)),           // Coral - House
    LoadingIcon(Icons.Default.Apartment, Color(0xFF6699FF)),       // Blue - Apartment
    LoadingIcon(Icons.Default.DirectionsCar, Color(0xFF4DCC80)),   // Green - Car
    LoadingIcon(Icons.Default.AirplanemodeActive, Color(0xFFFF9933)) // Orange - Airplane
)

/**
 * PlayStation-style loading animation with travel icons
 * Icons pulse and scale in sequence
 */
@Composable
fun MerryLoadingAnimation(
    modifier: Modifier = Modifier
) {
    val infiniteTransition = rememberInfiniteTransition(label = "loading")
    
    // Animate through 0-3 continuously
    val animatedIndex by infiniteTransition.animateFloat(
        initialValue = 0f,
        targetValue = 4f,
        animationSpec = infiniteRepeatable(
            animation = tween(1600, easing = LinearEasing),
            repeatMode = RepeatMode.Restart
        ),
        label = "index"
    )
    
    Column(
        modifier = modifier,
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Row(
            horizontalArrangement = Arrangement.spacedBy(20.dp),
            verticalAlignment = Alignment.CenterVertically
        ) {
            loadingIcons.forEachIndexed { index, item ->
                val isActive = animatedIndex.toInt() == index
                val scale = if (isActive) 1.3f else 1f
                val alpha = if (isActive) 1f else 0.4f
                
                Box(
                    modifier = Modifier.size(50.dp),
                    contentAlignment = Alignment.Center
                ) {
                    // Glow effect
                    if (isActive) {
                        Box(
                            modifier = Modifier
                                .size(50.dp)
                                .scale(scale)
                                .alpha(0.3f)
                                .background(item.color, RoundedCornerShape(25.dp))
                        )
                    }
                    
                    Icon(
                        imageVector = item.icon,
                        contentDescription = null,
                        modifier = Modifier
                            .size(28.dp)
                            .scale(scale)
                            .alpha(alpha),
                        tint = item.color
                    )
                }
            }
        }
        
        Spacer(modifier = Modifier.height(16.dp))
        
        Text(
            text = "Loading...",
            fontSize = 14.sp,
            color = Color.Gray
        )
    }
}

/**
 * Full screen loading overlay
 */
@Composable
fun MerryLoadingOverlay(
    message: String = "Loading...",
    modifier: Modifier = Modifier
) {
    Box(
        modifier = modifier
            .fillMaxSize()
            .background(Color.Black.copy(alpha = 0.4f)),
        contentAlignment = Alignment.Center
    ) {
        Surface(
            shape = RoundedCornerShape(20.dp),
            color = Color.White.copy(alpha = 0.95f),
            shadowElevation = 8.dp
        ) {
            Column(
                modifier = Modifier.padding(32.dp),
                horizontalAlignment = Alignment.CenterHorizontally
            ) {
                MerryLoadingAnimation()
                
                Spacer(modifier = Modifier.height(16.dp))
                
                Text(
                    text = message,
                    fontSize = 16.sp,
                    fontWeight = FontWeight.Medium,
                    color = Color(0xFF222222)
                )
            }
        }
    }
}

/**
 * Compact circular rotating loading animation
 */
@Composable
fun MerryLoadingCompact(
    modifier: Modifier = Modifier,
    size: Dp = 60.dp
) {
    val infiniteTransition = rememberInfiniteTransition(label = "rotation")
    
    val rotation by infiniteTransition.animateFloat(
        initialValue = 0f,
        targetValue = 360f,
        animationSpec = infiniteRepeatable(
            animation = tween(2000, easing = LinearEasing),
            repeatMode = RepeatMode.Restart
        ),
        label = "rotation"
    )
    
    Box(
        modifier = modifier.size(size),
        contentAlignment = Alignment.Center
    ) {
        loadingIcons.forEachIndexed { index, item ->
            val angle = Math.toRadians((index * 90 + rotation).toDouble())
            val radius = size.value / 3
            
            Icon(
                imageVector = item.icon,
                contentDescription = null,
                modifier = Modifier
                    .size(16.dp)
                    .offset(
                        x = (radius * cos(angle)).dp,
                        y = (radius * sin(angle)).dp
                    ),
                tint = item.color
            )
        }
    }
}

/**
 * Bouncing icons loading animation
 */
@Composable
fun MerryLoadingBounce(
    modifier: Modifier = Modifier
) {
    val infiniteTransition = rememberInfiniteTransition(label = "bounce")
    
    Row(
        modifier = modifier,
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.CenterVertically
    ) {
        loadingIcons.forEachIndexed { index, item ->
            val bounce by infiniteTransition.animateFloat(
                initialValue = 0f,
                targetValue = -15f,
                animationSpec = infiniteRepeatable(
                    animation = tween(
                        durationMillis = 500,
                        delayMillis = index * 150,
                        easing = FastOutSlowInEasing
                    ),
                    repeatMode = RepeatMode.Reverse
                ),
                label = "bounce$index"
            )
            
            Icon(
                imageVector = item.icon,
                contentDescription = null,
                modifier = Modifier
                    .size(24.dp)
                    .offset(y = bounce.dp),
                tint = item.color
            )
        }
    }
}

/**
 * Pulsing single icon loading
 */
@Composable
fun MerryLoadingPulse(
    modifier: Modifier = Modifier
) {
    val infiniteTransition = rememberInfiniteTransition(label = "pulse")
    
    val scale by infiniteTransition.animateFloat(
        initialValue = 0.8f,
        targetValue = 1.2f,
        animationSpec = infiniteRepeatable(
            animation = tween(800, easing = FastOutSlowInEasing),
            repeatMode = RepeatMode.Reverse
        ),
        label = "scale"
    )
    
    val alpha by infiniteTransition.animateFloat(
        initialValue = 0.5f,
        targetValue = 1f,
        animationSpec = infiniteRepeatable(
            animation = tween(800, easing = FastOutSlowInEasing),
            repeatMode = RepeatMode.Reverse
        ),
        label = "alpha"
    )
    
    Icon(
        imageVector = Icons.Default.Home,
        contentDescription = "Loading",
        modifier = modifier
            .size(48.dp)
            .scale(scale)
            .alpha(alpha),
        tint = Color(0xFFFF5744)
    )
}
